dmoj/cptbox/helper.cpp

```diff
@@ -175,6 +225,26 @@ int cptbox_child_run(const struct child_config *config) {
 #endif
 }
 
+int get_landlock_version() {
+#if !PTBOX_FREEBSD
+    char *sandbox_mode = getenv("DMOJ_SANDBOX_MODE");
+    if (sandbox_mode != nullptr && strcmp(sandbox_mode, "ptrace+seccomp") == 0) {
```

Comment:
This is a surprising place for this check to live, and I think the check should be stricter than this.

We should validate that the string is one of:

* `ptrace+seccomp`
* `ptrace+seccomp+landlock`
* `auto`

and otherwise bail out. Otherwise we can't confidently make changes to the interpretation of this environment variable, as we'd have been too lax in its inputs in earlier versions.

Reply:
I agree that this location is strange. Where should it live?

Reply:
I think living in this file is probably fine, and it should return an enum. Then it should be consulted in the spawn routine alongside the Landlock version we have available. My concern here is more about subtly overloading the semantics of `get_landlock_version` than anything else.

We should bail if someone requests:

* seccomp or seccomp+landlock support on FreeBSD
* ptrace-only support on Linux
* seccomp+landlock on a Linux with too old a Landlock ABI

This sanity check could exist in `judge.py` alongside all the other sanity checks, but I don't feel strongly about that.